File: src/alienclaw/api/audit_log.py
"""Submission audit log — append-only JSONL with daily rollover.

Storage layout:
    $ALIENCLAW_API_DATA_ROOT/audit/submissions-YYYY-MM-DD.jsonl

Each line (one JSON object):
    {
      "ts":             "<ISO 8601 UTC>",
      "api_key_hash":   "<sha256 hex — NEVER the raw key>",
      "client_ip":      "<IP string or 'unknown'>",
      "martian_type":   "<type>",
      "genome_sha256":  "<sha256 hex of submitted genome>",
      "fitness":        <float>,
      "result":         "accepted" | "rejected",
      "rejection_code": "<spec error code or null>"
    }

Write discipline:
- open() in append mode ('a') — POSIX O_APPEND guarantees atomic appends
  for writes smaller than PIPE_BUF (4096 bytes on Linux). Each log line
  is <512 bytes, well within the limit.
- fsync after each write to ensure the line reaches disk.
- Failures log to stderr and do NOT block the submission — audit failure
  is loud but non-blocking.

NEVER logs raw API keys. NEVER logs full genome content.
"""
from __future__ import annotations

import hashlib
import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


def _sha256(s: str) -> str:
    return hashlib.sha256(s.encode("utf-8")).hexdigest()


def _today_utc() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%d")


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


class AuditLog:
    """Append-only JSONL submission audit log."""

    def __init__(self, data_root: Path | None = None):
        self._root = data_root

    def _log_path(self, date_str: str | None = None) -> Path | None:
        if self._root is None:
            return None
        d = date_str or _today_utc()
        return self._root / "audit" / f"submissions-{d}.jsonl"

    def record(
        self,
        api_key_hash: str,
        martian_type: str,
        genome: str,
        fitness: float,
        result: str,
        rejection_code: str | None = None,
        client_ip: str = "unknown",
    ) -> None:
        """Append one log entry. Failures print to stderr; never raise."""
        path = self._log_path()
        if path is None:
            return

        entry: dict[str, Any] = {
            "ts":             _now_iso(),
            "api_key_hash":   api_key_hash,
            "client_ip":      client_ip or "unknown",
            "martian_type":   martian_type,
            "genome_sha256":  _sha256(genome),
            "fitness":        fitness,
            "result":         result,
            "rejection_code": rejection_code,
        }
        line = json.dumps(entry, sort_keys=True) + "\n"

        try:
            path.parent.mkdir(parents=True, exist_ok=True)
            with open(path, "a", encoding="utf-8") as f:
                f.write(line)
                f.flush()
                os.fsync(f.fileno())
        except Exception as exc:
            print(f"[audit_log] WARNING: failed to write entry: {exc}", file=sys.stderr)

File: src/alienclaw/api/audit-log.ts
/**
 * Submission audit log — append-only JSONL with daily rollover.
 * TypeScript port of api/audit_log.py (Packet 31.5).
 *
 * NEVER logs raw API keys or full genome content.
 */

import { createHash } from 'node:crypto';
import { appendFileSync, mkdirSync, fsyncSync, openSync, closeSync } from 'node:fs';
import { join } from 'node:path';

function sha256(s: string): string {
  return createHash('sha256').update(s, 'utf8').digest('hex');
}

function todayUtc(): string {
  return new Date().toISOString().slice(0, 10);
}

interface AuditEntry {
  ts:             string;
  api_key_hash:   string;
  client_ip:      string;
  martian_type:   string;
  genome_sha256:  string;
  fitness:        number;
  result:         'accepted' | 'rejected';
  rejection_code: string | null;
}

export class AuditLog {
  private readonly _root: string | null;

  constructor(opts: { dataRoot?: string } = {}) {
    this._root = opts.dataRoot ?? null;
  }

  private _logPath(date?: string): string | null {
    if (!this._root) return null;
    const d = date ?? todayUtc();
    return join(this._root, 'audit', `submissions-${d}.jsonl`);
  }

  record(opts: {
    apiKeyHash:     string;
    martianType:    string;
    genome:         string;
    fitness:        number;
    result:         'accepted' | 'rejected';
    rejectionCode?: string | null;
    clientIp?:      string;
  }): void {
    const path = this._logPath();
    if (!path) return;

    const entry: AuditEntry = {
      ts:             new Date().toISOString(),
      api_key_hash:   opts.apiKeyHash,
      client_ip:      opts.clientIp ?? 'unknown',
      martian_type:   opts.martianType,
      genome_sha256:  sha256(opts.genome),
      fitness:        opts.fitness,
      result:         opts.result,
      rejection_code: opts.rejectionCode ?? null,
    };
    const line = JSON.stringify(entry, Object.keys(entry).sort() as (keyof AuditEntry)[]) + '\n';

    try {
      const dir = path.slice(0, path.lastIndexOf('/'));
      mkdirSync(dir, { recursive: true });
      const fd = openSync(path, 'a');
      try {
        appendFileSync(fd, line, 'utf8');
        fsyncSync(fd);
      } finally {
        closeSync(fd);
      }
    } catch (err) {
      process.stderr.write(`[audit-log] WARNING: failed to write entry: ${err}\n`);
    }
  }
}
